### PR TITLE
feat(db): replace mysql with postgres

### PR DESCRIPTION
**Pull request by 6a68**
@pdehaan @nchapman @dannycoates r?

*edit: doh, I always think '@dcoates'. Thanks, Kit!*

### CODE REVIEW

File: server/bell_oauth_profile.js [hunk 1]
```diff
@@ -5,7 +5,7 @@
 'use strict';
 
 var log = require('./logger')('server.bell.profile');
-var db = require('./db/db');
+var user = require('./db/db')['user'];
```
Comment (pdehaan) [inline server/bell_oauth_profile.js]: Why this over `require('./db/db').user;`?

File: server/bell_oauth_profile.js [hunk 2]
```diff
@@ -17,16 +17,16 @@ function profile (credentials, params, get, profileCb) {
     // Bell returns the parsed data and handles errors internally
     log.verbose('exchanged tokens for profile data:' + JSON.stringify(data));
     // TODO use Joi to validate `data` before sending to DB
-    db.getUserById(data.uid, function (err, user) {
+    user.get(data.uid, function (err, user) {
       if (err) {
-        log.warn('getUserById failed: ' + err);
-        return profileCb('getUserById failed: ' + err);
+        log.warn('user.get failed: ' + err);
+        return profileCb('user.get failed: ' + err);
```
Comment (pdehaan) [inline server/bell_oauth_profile.js]: What's the signature for `profileCb()`? Should this just be returning a `profileCb(err)`?

File: server/config.js [hunk 3]
```diff
@@ -49,38 +49,38 @@ var conf = convict({
...
+        env: 'PGHOST',
+        format: String
       },
       port: {
-        default: 3306,
-        env: 'MYSQL_PORT',
-        format: 'port'
+        default: 5432,
+        env: 'PGPORT',
+        format: 'int'
```
Comment (pdehaan) [inline server/config.js]: `'int'` or `'port'`?

File: server/db/create_db.sh [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+  url TEXT NOT NULL,
+  rawUrl TEXT NOT NULL,
+  urlHash CHAR(40) NOT NULL,
+  title TEXT NOT NULL,
+  visitedAt TIMESTAMPTZ(3) NOT NULL,
+  updatedAt TIMESTAMPTZ(3)
+);" -d chronicle -U chronicle
+psql -c "CREATE UNIQUE INDEX CONCURRENTLY fxaId_visitedAt_id
+  ON visits (fxaId, visitedAt, id);" -d chronicle -U chronicle
+
```
Comment (pdehaan) [inline server/db/create_db.sh]: Stupid idea, but was wondering if we could convert this to `#!/bin/env node` and then just use the convict vars and a bunch of shell `exec()` calls. But maybe that'd be too gross for words.

Or maybe even if we could pass in some ENV vars, so we could do `$ DATABASE=chronicle USER=chronicle server/db/create_db.sh` and then use the env vars in the script.

File: server/db/create_test_data.js [hunk 5]
```diff
@@ -4,14 +4,19 @@
 
 // this script assumes you've run the create_db script.
 
+//jscs:disable
```
Comment (pdehaan) [inline server/db/create_test_data.js]: Feelings GIF:

File: server/db/create_test_data.js [hunk 6]
```diff
@@ -4,14 +4,19 @@
...
 
 var createHash = require('crypto').createHash;
 var config = require('../config');
-var db = require('./db');
+var models = require('./db');
+var user = models.user;
+var visit = models.visit;
 var log = require('../logger')('db.createTestUser');
 
 var pool = require('./utils').createPool();
```
Comment (pdehaan) [inline server/db/create_test_data.js]: Do we still have a `createPool()`?

File: server/db/create_test_data.js [hunk 7]
```diff
@@ -58,45 +72,43 @@ function createTestUser(recordCount, cb) {
...
+          done();
+          pg.end();
+          return cb && cb(err);
         }
-        c.release();
+        // TODO examine r and get the number of affected rows
+        log.verbose('inserted ' + recordCount + ' user visits into db');
         cb && cb(err);
+        done();
+        pg.end();
```
Comment (pdehaan) [inline server/db/create_test_data.js]: Since we're always calling `done()` and `pg.end()`, does it make sense to move those into the callback? Or doesn't that make sense since we seem to not always be guaranteed to have a callback.
Conversely, if no callback function is specified, create a dummy function which calls `pg.end()`.

File: server/db/create_test_data.js [hunk 8]
```diff
@@ -58,45 +72,43 @@ function createTestUser(recordCount, cb) {
...
+      /*jshint quotmark: false */
+      while(bulkData.length) {
+        var next = bulkData.shift();
+        query += '(' + next.map(function(item) { return "'" + item + "'"; }).join(', ') + ')';
+        if (bulkData.length) { query += ', '; }
       }
-      c.query('TRUNCATE users', function(err) {
+      log.trace('about to insert bulk data');
+      log.trace('query is: ' + query);
+      log.trace('bulkData.length is ' + bulkData.length);
```
Comment (pdehaan) [inline server/db/create_test_data.js]: Wouldn't `bulkData.length` always be 0 here if we're using a `while (bulkData.length)` and `bulkData.shift()`?

File: server/db/create_test_data.js [hunk 9]
```diff
@@ -116,9 +128,4 @@ program
   .parse(process.argv);
 
 var count = program.count || defaultCount;
-truncateTables(function(err) {
-  if (err) {
-    throw err;
-  }
-  createTestUser.call(null, count, done);
-});
+createTestUser(count, done);
```
Comment (pdehaan) [inline server/db/create_test_data.js]: This could probably just be `createTestUser(program.count)` since the `createTestUser()` func already defaults to a `defaultCount` (see line 43).

File: server/db/create_test_data.js [hunk 10]
```diff
@@ -4,14 +4,19 @@
...
 var program = require('commander');
+var pg = require('pg');
 
 var createHash = require('crypto').createHash;
 var config = require('../config');
-var db = require('./db');
+var models = require('./db');
+var user = models.user;
+var visit = models.visit;
 var log = require('../logger')('db.createTestUser');
```
Comment (pdehaan) [inline server/db/create_test_data.js]: This should be `('db.createTestData')`, no?

File: server/db/db.js [hunk 11]
```diff
@@ -4,123 +4,190 @@
...
-        log.warn('error getting visit: ' + err);
-      }
-      cb(err, r && r[0]);
+                'FROM visits WHERE id = $1 AND fxaId = $2';
+    pg.connect(dbParams, function(err, client, done) {
+      client.query(query, [visitId, fxaId], function(err, r) {
+        if (err) {
+          log.warn('error getting visit: ' + err);
+        } else {
+          _verbose('db.visits.get succeeded');
```
Comment (pdehaan) [inline server/db/db.js]: `'db.visit.get succeeded'` (singular "visit", to match L122 above)?

File: server/db/db.js [hunk 12]
```diff
@@ -4,123 +4,190 @@
...
-        log.warn('error creating visit: ' + err);
-      }
-      cb(err);
+                'VALUES ($1, $2, $3, $4, $5, $6, $7)';
+    pg.connect(dbParams, function(err, client, done) {
+      client.query(query, [visitId, fxaId, url, url, urlHash, title, visitedAt], function(err, r) {
+        if (err) {
+          log.warn('error creating visit: ' + err);
+        } else {
+          _verbose('db.visits.create succeeded');
```
Comment (pdehaan) [inline server/db/db.js]: "visit" singular?

File: server/db/db.js [hunk 13]
```diff
@@ -4,123 +4,190 @@
...
-      cb(err);
+  delete: function(fxaId, visitId, cb) {
+    _verbose('db.visit.delete', fxaId, visitId);
+    var query = 'DELETE FROM visits WHERE fxaId = $1 AND id = $2';
+    pg.connect(dbParams, function(err, client, done) {
+      client.query(query, [fxaId, visitId], function(err) {
+        if (err) {
+          log.warn('error deleting visit: ' + err);
+        } else {
+          _verbose('db.visits.delete succeeded');
```
Comment (pdehaan) [inline server/db/db.js]: "visit" singular?

File: server/index.js [hunk 14]
```diff
@@ -7,11 +7,13 @@
...
 var HapiAuthCookie = require('hapi-auth-cookie');
 var Bell = require('bell');
+var pg = require('pg');
+
 var config = require('./config');
 var log = require('./logger')('server.index');
 var authProfileCb = require('./bell_oauth_profile');
 var routes = require('./routes');
-var db = require('./db/db');
+var user = require('./db/db')['user'];
```
Comment (pdehaan) [inline server/index.js]: Why not just `require('./db/db').user;`?

File: server/index.js [hunk 15]
```diff
@@ -35,11 +37,11 @@ server.connection({
 
 server.register([
   HapiAuthCookie,
-  Bell
+  Bell,
```
Comment (pdehaan) [inline server/index.js]: Trailing comma alert!

File: README.md [hunk 16]
```diff
@@ -25,8 +25,11 @@ To fetch dependencies and get cooking:
 1. `npm install`
 2. As part of the npm install process, the `postinstall` script will install the Bower dependencies for you.
 3. Copy `config/local.json.example` to `config/local.json`, and put your local info in there.
-4. `npm start`
-5. You're up and running! surf to <http://localhost:8080> :surfer:
+4. Run `./server/db/create_db.sh` to create the database, tables, and indexes
```
Comment (pdehaan) [inline README.md]: Dumb question, but should this be in `server/db/create_db.sh` or just `bin/create_db.sh` (along w/ `bin/www`)?

File: server/bell_oauth_profile.js [hunk 17]
```diff
@@ -36,7 +36,7 @@ function profile (credentials, params, get, profileCb) {
         });
       } else {
         log.verbose('new user! creating record and setting session cookie...');
-        db.createUser(data.uid, data.email, params.access_token, function(err) {
+        user.create(data.uid, data.email, params.access_token, function(err) {
           if (err) {
             log.warn('user creation failed: ' + err);
```
Comment (pdehaan) [inline server/bell_oauth_profile.js]: This should this be `log.warn('user.create failed: ' + err);` for consistency?

File: server/db/create_db.sh [hunk 18]
```diff
@@ -0,0 +1,35 @@
...
+psql -c "CREATE USER chronicle WITH PASSWORD 'chronicle';" -U postgres
+psql -c "CREATE DATABASE chronicle ENCODING 'UTF-8' LC_COLLATE = 'en_US.UTF-8' LC_CTYPE = 'en_US.UTF-8';" -U postgres
+psql -c 'GRANT ALL PRIVILEGES ON DATABASE chronicle to chronicle;' -U postgres
+psql -c 'ALTER SCHEMA public OWNER TO chronicle;' -U postgres
+psql -c "CREATE TABLE IF NOT EXISTS users (
+  fxaId CHAR(32) PRIMARY KEY,
+  email VARCHAR(255) NOT NULL,
+  oauthToken TEXT,
+  createdAt TIMESTAMPTZ(3) NOT NULL,
+  updatedAt TIMESTAMPTZ(3)
```
Comment (pdehaan) [inline server/db/create_db.sh]: Should `createdAt` and `updatedAt` default to the current time? (same w/ visits table down there in lines 30-31)

File: server/db/create_test_data.js [hunk 19]
```diff
@@ -4,14 +4,19 @@
...
 var uuid = require('uuid');
 var program = require('commander');
+var pg = require('pg');
 
 var createHash = require('crypto').createHash;
 var config = require('../config');
-var db = require('./db');
+var models = require('./db');
+var user = models.user;
+var visit = models.visit;
```
Comment (pdehaan) [inline server/db/create_test_data.js]: `visit` is defined but never used

File: server/db/create_test_data.js (hunk 6 above)
Comment (pdehaan) [inline server/db/create_test_data.js]: `pool` is defined but never used

File: server/db/create_test_data.js [hunk 20]
```diff
@@ -58,45 +72,43 @@ function createTestUser(recordCount, cb) {
...
+      while(bulkData.length) {
+        var next = bulkData.shift();
+        query += '(' + next.map(function(item) { return "'" + item + "'"; }).join(', ') + ')';
+        if (bulkData.length) { query += ', '; }
       }
-      c.query('TRUNCATE users', function(err) {
+      log.trace('about to insert bulk data');
+      log.trace('query is: ' + query);
+      log.trace('bulkData.length is ' + bulkData.length);
+      client.query(query, function(err, r) {
```
Comment (pdehaan) [inline server/db/create_test_data.js]: `r` is defined but never used

File: server/db/db.js [hunk 21]
```diff
@@ -4,123 +4,190 @@
...
 'use strict';
 
-var createHash = require('crypto').createHash;
-var log = require('../logger')('server.db.db');
-var utils = require('./utils');
+var crypto = require('crypto');
 
-var pool = utils.createPool();
+var pg = require('pg');
+var uuid = require('uuid');
```
Comment (pdehaan) [inline server/db/db.js]: `uuid` is defined but never used

File: server/db/db.js [hunk 22]
```diff
@@ -4,123 +4,190 @@
...
-      _verbose('db.createUser: created user');
-      cb(err);
+// TODO if we get an unrecoverable error from a connection, we want to kill it,
+// so that it is not returned to the pool (hapi-node-postgres suggestion).
+var user = {
+  create: function(fxaId, email, oauthToken, cb) {
+    _verbose('db.user.create', fxaId, email, oauthToken);
+    var query = 'INSERT INTO users (fxaId, email, oauthToken, createdAt) ' +
+                'VALUES ($1, $2, $3, $4)';
+    pg.connect(dbParams, function(err, client, done) {
```
Comment (pdehaan) [inline server/db/db.js]: We aren't handing `err` here. We may want to fail fast-n-hard.

File: server/db/db.js [hunk 23]
```diff
@@ -4,123 +4,190 @@
...
-    var query = 'SELECT fxaId, email, oauthToken FROM users WHERE fxaId = ?';
-    pool.query(query, fxaId, function(err, r) {
-      if (err) {
-        log.warn('error retrieving user: ' + err);
-      }
-      cb(err, r && r[0]);
+  get: function(fxaId, cb) {
+    _verbose('db.user.get called', fxaId);
+    var query = 'SELECT fxaId, email FROM users WHERE fxaId = $1';
+    pg.connect(dbParams, function(err, client, done) {
```
Comment (pdehaan) [inline server/db/db.js]: unused `err`.

File: server/db/db.js [hunk 24]
```diff
@@ -4,123 +4,190 @@
...
-        // TODO send the item to a retry queue?
-        return cb(err);
-      }
-      _verbose('db.updateUser: updated user');
-      cb(err);
+  update: function(fxaId, email, oauthToken, cb) {
+    _verbose('db.user.update', fxaId, email, oauthToken);
+    var query = 'UPDATE users SET email = $1, oauthToken = $2, updatedAt = $3 ' +
+                'WHERE fxaId = $4 RETURNING email, fxaId';
+    pg.connect(dbParams, function(err, client, done) {
```
Comment (pdehaan) [inline server/db/db.js]: unused `err`.

File: server/db/db.js [hunk 25]
```diff
@@ -4,123 +4,190 @@
...
     var query = 'SELECT id, url, urlHash, title, visitedAt ' +
-                'FROM visits WHERE fxaId = ? ' +
-                'AND visitedAt < (SELECT visitedAt FROM visits WHERE id = ?) ' +
-                'ORDER BY visitedAt DESC LIMIT ?';
-    pool.query(query, [fxaId, visitId, count], function(err, results) {
-      cb(err, results);
+                'FROM visits WHERE fxaId = $1 ' +
+                'AND visitedAt < (SELECT visitedAt FROM visits WHERE id = $2) ' +
+                'ORDER BY visitedAt DESC LIMIT $3';
+    pg.connect(dbParams, function(err, client, done) {
```
Comment (pdehaan) [inline server/db/db.js]: unused `err`.

File: server/db/db.js [hunk 26]
```diff
@@ -4,123 +4,190 @@
...
-  getVisits: function(fxaId, count, cb) {
-    _verbose('db.getVisits', arguments);
+  get: function(fxaId, count, cb) {
+    _verbose('db.visits.get', fxaId, count);
     var query = 'SELECT id, url, urlHash, title, visitedAt ' +
-                'FROM visits WHERE fxaId = ? ORDER BY visitedAt DESC LIMIT ?';
-    pool.query(query, [fxaId, count], function(err, results) {
-      cb(err, results);
+                'FROM visits WHERE fxaId = $1 ORDER BY visitedAt DESC LIMIT $2';
+    pg.connect(dbParams, function(err, client, done) {
```
Comment (pdehaan) [inline server/db/db.js]: unused `err`.

File: server/db/db.js [hunk 27]
```diff
@@ -4,123 +4,190 @@
...
+    // important: postgres enforces that the user with id `fxaId` is the user with visit `visitId`
     var query = 'SELECT id, url, urlHash, title, visitedAt ' +
-                'FROM visits WHERE id = ? AND fxaId = ?';
-    pool.query(query, [visitId, fxaId], function(err, r) {
-      if (err) {
-        log.warn('error getting visit: ' + err);
-      }
-      cb(err, r && r[0]);
+                'FROM visits WHERE id = $1 AND fxaId = $2';
+    pg.connect(dbParams, function(err, client, done) {
```
Comment (pdehaan) [inline server/db/db.js]: unused `err`.

File: server/db/db.js [hunk 28]
```diff
@@ -4,123 +4,190 @@
...
+    var urlHash = crypto.createHash('sha1').update(url).digest('hex').toString();
     var query = 'INSERT INTO visits (id, fxaId, rawUrl, url, urlHash, title, visitedAt) ' +
-                'VALUES (?, ?, ?, ?, ?, ?, ?)';
-    pool.query(query, [visitId, fxaId, url, url, urlHash, title, visitedAt], function(err, results) {
-      if (err) {
-        log.warn('error creating visit: ' + err);
-      }
-      cb(err);
+                'VALUES ($1, $2, $3, $4, $5, $6, $7)';
+    pg.connect(dbParams, function(err, client, done) {
```
Comment (pdehaan) [inline server/db/db.js]: unused `err`.

File: server/db/db.js [hunk 29]
```diff
@@ -4,123 +4,190 @@
...
-        log.warn('error updating visit: ' + err);
-      }
-      cb(err, r && r[0]);
+  update: function(fxaId, visitId, visitedAt, url, title, cb) {
+    _verbose('db.visit.update', fxaId, visitId, visitedAt, url, title);
+    var query = 'UPDATE visits SET visitedAt = $1, updatedAt = $2, url = $3, ' +
+                'urlHash = $4, rawUrl = $5, title = $6 ' +
+                'WHERE fxaId = $7 AND id = $8 RETURNING id, fxaId, visitedAt, url, urlHash, title';
+    var urlHash = crypto.createHash('sha1').update(url).digest('hex').toString();
+    pg.connect(dbParams, function(err, client, done) {
```
Comment (pdehaan) [inline server/db/db.js]: unused `err`.

File: server/db/db.js [hunk 30]
```diff
@@ -4,123 +4,190 @@
...
-    var query = 'DELETE FROM visits WHERE fxaId = ? AND id = ?';
-    pool.query(query, [fxaId, visitId], function(err) {
-      if (err) {
-        log.warn('error deleting visit: ' + err);
-      }
-      cb(err);
+  delete: function(fxaId, visitId, cb) {
+    _verbose('db.visit.delete', fxaId, visitId);
+    var query = 'DELETE FROM visits WHERE fxaId = $1 AND id = $2';
+    pg.connect(dbParams, function(err, client, done) {
```
Comment (pdehaan) [inline server/db/db.js]: unused `err`.

File: server/routes/visits.js [hunk 31]
```diff
@@ -87,7 +89,7 @@ module.exports = [{
       var p = request.payload;
       var fxaId = request.auth.credentials;
       var visitId = p.visitId || uuid.v4();
-      db.createVisit(fxaId, visitId, p.visitedAt, p.url, p.title, function (err, visit) {
+      visit.create(fxaId, visitId, p.visitedAt, p.url, p.title, function (err, visit) {
```
Comment (pdehaan) [inline server/routes/visits.js]: `visit` is defined but never used.

Not sure if the return value name of `visit` is confusing since we're already using `visit` as a global alias for the `models.visit`.

File: README.md (hunk 16 above)
Comment (pdehaan) [inline README.md]: Hhmmm,  I'm getting this error when trying to run this in mein Terminal:

```sh
$ ./server/db/create_db.sh

psql: could not connect to server: No such file or directory
	Is the server running locally and accepting
	connections on Unix domain socket "/tmp/.s.PGSQL.5432"?
...
```

Comment (pdehaan) [inline README.md]: Actually, even after brew uninstalling postgresql (9.3) and reinstalling (9.4), I'm seeing this:

```sh
$ psql
psql: could not connect to server: No such file or directory
	Is the server running locally and accepting
	connections on Unix domain socket "/tmp/.s.PGSQL.5432"?

$ psql --version
psql (PostgreSQL) 9.4.0
```

I must go deeper.

**Comment by 6a68** [comment]
@pdehaan you are my review hero

**Comment by 6a68** [comment]
Turns out you can't put gifs in a git commit, oh well

File: README.md (hunk 16 above)
Comment (6a68) [inline README.md]: yikes, that's...



srsly, give it a shot with a fresh 9.3.5 install (I think it's still stable? and that's the latest RDS version), paste your results in here, we can try to debug further

File: server/db/create_db.sh (hunk 18 above)
Comment (6a68) [inline server/db/create_db.sh]: nah, in postgres that would have to be done with a trigger, i.e., funky dba-styley programming. we can move the default timestamp insertion into the app instead, in the spirit of remaining database agnostic

File: server/db/db.js (hunk 29 above)
Comment (6a68) [inline server/db/db.js]: am I stuck in an infinite scroll loop rn

File: server/db/db.js (hunk 30 above)
Comment (6a68) [inline server/db/db.js]: oh good never mind

File: server/db/create_db.sh [hunk 32]
```diff
@@ -0,0 +1,35 @@
+#!/bin/bash
+
+# This Source Code Form is subject to the terms of the Mozilla Public
+# License, v. 2.0. If a copy of the MPL was not distributed with this
+# file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+# TODO convert to node when things slow down
+psql -c 'DROP DATABASE IF EXISTS chronicle;' -U postgres
```
Comment (nchapman) [inline server/db/create_db.sh]: I get an error here because I don't have a `postgres` role by default.

File: server/bell_oauth_profile.js (hunk 1 above)
Comment (6a68) [inline server/bell_oauth_profile.js]: I...I don't know. :blush:

**Comment by dannycoates** [comment]
pdlint ftw! 

I'm just now taking a look.

File: server/bell_oauth_profile.js (hunk 2 above)
Comment (6a68) [inline server/bell_oauth_profile.js]: Hmm, good question. The signature is `profile(err, credentials)`, according to the [bell docs](https://github.com/hapijs/bell). However, looking at the code where [bell calls this function](https://github.com/hapijs/bell/blob/master/lib/oauth.js#L257-L260), it actually doesn't check for an error at all. Kinda funky. This `profile` function doesn't have access to the `reply` or `response` functions...maybe the right thing is to emit an error directly? :-\

File: server/config.js (hunk 3 above)
Comment (6a68) [inline server/config.js]: +1

**Comment by pdehaan** [comment]
Bonus points if somebody wants to write a GitHub hook for all our repos that will detect when we summon the wrong @dannycoates GitHub acct in a comment and auto-correct it.

File: server/db/create_db.sh (hunk 4 above)
Comment (6a68) [inline server/db/create_db.sh]: Good idea! I also considered using mustache to template in the values, maybe as part of the build process...probably just going to leave this as-is, but PRs very welcome as a follow-on fix.

File: server/db/create_test_data.js (hunk 5 above)
Comment (6a68) [inline server/db/create_test_data.js]: lol, I couldn't figure out how to get jscs to ignore the double quotes. I'll fix.

File: server/db/create_test_data.js (hunk 6 above)
Comment (6a68) [inline server/db/create_test_data.js]: good catch

File: server/db/create_test_data.js (hunk 5 above)
Comment (pdehaan) [inline server/db/create_test_data.js]: _double quotes_? You **ANIMAL**!

Comment (6a68) [inline server/db/create_test_data.js]: yeah, because line 94

File: server/db/create_db.sh (hunk 32 above)
Comment (6a68) [inline server/db/create_db.sh]: :-( I, um, thought that role should always exist? I guess I can just not specify -U and assume that the `$USER` superuser role is available?

Comment (6a68) [inline server/db/create_db.sh]: Pretty weird how different our homebrew-installed versions of postgres are, considering the only difference is that I'm on mavericks.

Comment (nchapman) [inline server/db/create_db.sh]: Yeah I don't get it either. That role definitely exists on Travis. I'll look into it on my end.

File: server/db/create_test_data.js (hunk 5 above)
Comment (pdehaan) [inline server/db/create_test_data.js]: Ah yes... I remember that line. That would be much easier w/ Knex.

How about jus' escapin' the single quote:
```js
query += '(' + next.map(function(item) { return '\'' + item + '\''; }).join(', ') + ')';
```

File: server/db/create_db.sh (hunk 4 above)
Comment (pdehaan) [inline server/db/create_db.sh]: Interesting idea... IIRC, @nchapman was using some grunt-template module (see [/grunttasks/template.js](https://github.com/mozilla/chronicle/blob/master/grunttasks/template.js)) to handle some template vars in some config file... Not too sure of the details, but I think he's just doing some string substitutions on this file: [/app/scripts/config.js](https://github.com/mozilla/chronicle/blob/master/app/scripts/config.js). Maybe something similar could be hacked for create_db.sh using grunt-template (if the create_db.js Node script idea doesnt fly).

File: server/db/db.js [hunk 33]
```diff
@@ -4,123 +4,190 @@
...
     var query = 'INSERT INTO visits (id, fxaId, rawUrl, url, urlHash, title, visitedAt) ' +
-                'VALUES (?, ?, ?, ?, ?, ?, ?)';
-    pool.query(query, [visitId, fxaId, url, url, urlHash, title, visitedAt], function(err, results) {
-      if (err) {
-        log.warn('error creating visit: ' + err);
-      }
-      cb(err);
+                'VALUES ($1, $2, $3, $4, $5, $6, $7)';
+    pg.connect(dbParams, function(err, client, done) {
+      client.query(query, [visitId, fxaId, url, url, urlHash, title, visitedAt], function(err, r) {
```
Comment (pdehaan) [inline server/db/db.js]: So, we're inserting the URL here as both `rawUrl` and `url`. Do we want to be normalizing one of those or something?
I saw this via npm tweets the other day, not sure if it's useful: https://github.com/sindresorhus/normalize-url

Or, if we want to wrap our own util function around URL normalization &mdash; so we can strip out certain URL query params (e.g. Google analytics tracking codes, etc).

Or, tell me to shut up and we can kick this can far, far down the road.

**Comment by dannycoates** [comment]
I had a bit of trouble getting `create_db.sh` working on an ubuntu 14.04 vm with a vanilla `apt-get install postgresql`. First I had to use `sudo -u postgres ./create_db.sh` and then I had to add `-h localhost` to the lines that used the chronicle user, which then had me type in the password each time. Eventually the whole script ran.

This is literally the first time I've used pg, so I have no idea what I'm doing.

File: server/db/db.js [hunk 34]
```diff
@@ -4,123 +4,190 @@
...
+  create: function(fxaId, email, oauthToken, cb) {
+    _verbose('db.user.create', fxaId, email, oauthToken);
+    var query = 'INSERT INTO users (fxaId, email, oauthToken, createdAt) ' +
+                'VALUES ($1, $2, $3, $4)';
+    pg.connect(dbParams, function(err, client, done) {
+      client.query(query, [fxaId, email, oauthToken, new Date().toJSON()], function(err) {
+        if (err) {
+          log.warn('error saving user: ' + err);
+          // TODO send the item to a retry queue?
+          return cb(err);
```
Comment (dannycoates) [inline server/db/db.js]: this return skips `done()` prob just delete it like the rest

**Comment by dannycoates** [comment]
I get `Error: no test specified` when I run the tests :trollface:

Anyway, this all looks reasonable to me. I found a couple things while setting up my environment that I'll post separate issues for but otherwise, LGTM.

File: server/db/create_db.sh (hunk 4 above)
Comment (6a68) [inline server/db/create_db.sh]: opened #116 to track this task

File: server/db/create_test_data.js (hunk 5 above)
Comment (6a68) [inline server/db/create_test_data.js]: good idea! it's pretty much unreadable code either way

File: server/db/create_test_data.js (hunk 7 above)
Comment (6a68) [inline server/db/create_test_data.js]: that sounds yummy

File: server/db/create_test_data.js (hunk 8 above)
Comment (6a68) [inline server/db/create_test_data.js]: doh

File: server/db/db.js (hunk 11 above)
Comment (6a68) [inline server/db/db.js]: ++

File: server/db/create_test_data.js (hunk 10 above)
Comment (6a68) [inline server/db/create_test_data.js]: aye

File: server/db/db.js (hunk 12 above)
Comment (6a68) [inline server/db/db.js]: aye

File: server/db/db.js (hunk 13 above)
Comment (6a68) [inline server/db/db.js]: yarrrr

File: server/index.js (hunk 15 above)
Comment (6a68) [inline server/index.js]: ;-)

File: server/db/db.js (hunk 21 above)
Comment (6a68) [inline server/db/db.js]: that actually was a really bad goof; visitId is optional, and the code should insert a `uuid.v4()` if the arg is missing

File: server/db/db.js (hunk 22 above)
Comment (6a68) [inline server/db/db.js]: added a generic handler to all the pg.connects.

File: server/routes/visits.js (hunk 31 above)
Comment (6a68) [inline server/routes/visits.js]: good catch! this should have been `reply(result)`, like with the `PUT` response

File: server/db/db.js (hunk 33 above)
Comment (6a68) [inline server/db/db.js]: shame cube engage

just kiddin', the idea here is to store `url` and `rawUrl` so that we can iteratively improve our normalization/canonicalization code; we need the originals to take more than one pass. Right now I'm explicitly ducking the normalization problem (see the wiki page related to this); we'll probably wind up using the normalized URL returned by embedly for a first pass at the fancy `url`.

Comment (6a68) [inline server/db/db.js]: clarifying slightly further: right now url == rawUrl because I'm just plain not cleaning up the URLs at all. But that should change any minute now when we add embedly. Why don't I add this to the scraper issue as a reminder :-)

File: server/db/db.js (hunk 34 above)
Comment (6a68) [inline server/db/db.js]: awesome! thanks

**Comment by 6a68** [comment]
yeah, we don't have any tests right now; the plan is to use `intern` for light functional testing, but mainly try to trade quality for speed

**Comment by 6a68** [comment]
excellent work team! merging

File: README.md (hunk 16 above)
Comment (pdehaan) [inline README.md]: OK, I think this is getting me closer (to go full circle). I was getting this error:
```sh
$ postgres
LOG:  skipping missing configuration file "/usr/local/var/postgres/postgresql.auto.conf"
FATAL:  database files are incompatible with server
DETAIL:  The data directory was initialized by PostgreSQL version 9.3, which is not compatible with this version 9.4.0.
```
This seemed to help me reach database nirvana: https://github.com/Homebrew/homebrew/issues/35240

TL;DR:
```sh
$ rm -rf /usr/local/var/postgres/
$ initdb /usr/local/var/postgres/
```

Comment (6a68) [inline README.md]: aha, so it was a homebrew bug!
